adding indexing / index advisor tutorial to user profile tutorial (#8)
couchbaselabs/tutorials-contrib: modules/profile-store/pages/index.adoc

> :imagesdir: ../assets/images

Reviewer: ```suggestion
```
We usually keep attributes after the h1 level heading

> = Building the right indexes for querying a User Profile Store

Reviewer: I've been missing some context here. We usually reserve **index.adoc** for introduction/set-up instruction files, for this tutorial it's already the install page https://docs.couchbase.com/tutorials/profile-store/install.html

Suggest renaming this file to **building-indexes.adoc** or something similar to avoid the possible confusion.

> = Building the right indexes for querying a User Profile Store

Reviewer: ```suggestion
= Building the right indexes for querying a User Profile Store
:imagesdir: ../assets/images
```

== Overview

> This tuturial is part of the main User Profile tutorial (both link:http://docs.couchbase.com/tutorials/profile-store/dotnet.html[.NET] and link:http://docs.couchbase.com/tutorials/profile-store/java.html[Java]). You should be able to follow along if you haven't gone through those tutorials yet, but reading those tutorials first will give you more context.

Reviewer: ```suggestion
This tutorial is part of the main User Profile tutorial (both xref:dotnet.adoc[.NET] and xref:java.adoc[Java]). You should be able to follow along if you haven't gone through those tutorials yet, but reading those tutorials first will give you more context.
```

* <<Improved Performance>> We'll analyze the improvement when using a better index.
* <<Summary>> A quick summary of this tutorial and next steps.

== Loading User Profile Data

If you've been following along with the user profile tutorial, your Couchbase bucket will have at least one document that looks like this:

[source,JavaScript,indent=0]
----
{
  "firstName": "Denis",
  "lastName": "Rosa",
  "enabled": true,
  "tenantId" : 4,
  "countryCode": "DE",
  "username": "someUser",
  "password": "letMeIn",
  "securityRoles": ["USER", "ADMIN"],
  "preferences": [
    {
      "name": "lang",
      "value": "en"
    },
    {
      "name": "currency",
      "value": "EUR"
    }
  ],
  "addresses": [
    {
      "name": "Home",
      "street": "Marienplatz",
      "number": "1",
      "zipCode": "80913",
      "city": "Munich",
      "state": "Bayern",
      "countryCode": "DE"
    }
  ],
  "telephones": [
      {
          "name": "cell",
          "number": "111-222-3333"
      }
  ]
}
----

You can use those documents for this tutorial. Alternatively, you can import a sample data set that's been created for this tutorial.

> Start by link:https://github.com/couchbaselabs/tutorials-contrib/blob/master/modules/profile-store/examples/user-profile-sample.json?raw=true[downloading the user-profile-sample.json file (large JSON file)].

Reviewer: ```suggestion
Start by https://github.com/couchbaselabs/tutorials-contrib/blob/master/modules/profile-store/examples/user-profile-sample.json?raw=true[downloading the user-profile-sample.json file (large JSON file)].
```

[NOTE]
====

> * This file should be saved to some location where you have cbimport installed or that can be accessed directly by a computer with cbimport installed. (cbimport comes with Couchbase Server)

Reviewer: ```suggestion
* This file should be saved to some location where you have cbimport installed or that can be accessed directly by a computer with cbimport installed. (cbimport comes with Couchbase Server).
```

* The computer with the file must have network access to the Couchbase cluster you want to load the data into. If you've been running a local instance of Couchbase or a local Docker image, download this file to the same computer. Otherwise, you may need to adjust firewall settings.

* Finally, make sure you have a bucket ready in Couchbase. For the user profile tutorial, the bucket was called *user_profile*, so that's what this tutorial will also be using.
====

> Review the cbimport documentation, specifically the link:https://docs.couchbase.com/server/current/tools/cbimport-json.html[cbimport json] documentation.

Reviewer: ```suggestion
Review the cbimport documentation, specifically the https://docs.couchbase.com/server/current/tools/cbimport-json.html[cbimport json] documentation.
```

To import the data into the user_profile bucket, execute a command like this:

[source,PowerShell,indent=0]
----
cbimport json -c localhost -u Administrator -p
password -b user_profile -d file://user-profile-sample.json -f list -g %id%
----

After this is complete, you should see a message like this:

[source,PowerShell,indent=0]
----
Json `file://user-profile-sample.json` imported to `http://localhost:8091` successfully
Documents imported: 25000 Documents failed: 0
----

> You should have (at least) 25000 records in your user_profile bucket.

Author: Yes, it should be in the nav, but it should be in the nav for the #9 PR, it's not part of the user profile tutorial

[source,SQL,indent=0]
----
SELECT META().id as id, username, tenantId, firstName, lastname
FROM user_profile
WHERE type = 'user'
AND tenantId = 5
ORDER BY firstName ASC
LIMIT 50
OFFSET 250
----

This query will take a long time to execute. Make a benchmark of the execution time (e.g. 39.6 seconds).

image:00305-execution-time.png[Execution time using a primary index]

If you have Couchbase Server 6.5 beta (or you've enabled developer preview features in Couchbase Server 6.5), you should notice an "Advise" button.

NOTE: Query Advisor is only available starting in Couchbase Server 6.5 Beta, as a preview feature. As this tutorial was written, it is cutting edge. It is a very useful tool already, but please use at your own risk.

Click the "Advise" button to get advise about indexing that will improve execution time:

image:00306-advise-results.png[Advise results]

The advisor will show what index is currently being used (a primary index) and what index it recommends. It won't always have a suggestion, but in this case, it recommends an index:

[source,SQL,indent=0]
----
CREATE INDEX adv_tenantId_type ON `user_profile`(`tenantId`) WHERE `type` = 'user'
----

> A quick sanity check shows that this suggestion makes sense. Our query is always examing `tenantId` and is always looking at `user` documents.

Reviewer: ```suggestion
A quick sanity check shows that this suggestion makes sense. Our query is always examining `tenantId` and is always looking at `user` documents.
```

== Summary

In this mini-tutorial, we looked at how to go from a query using an inefficient primary index to using index advisor to create a more efficient index.

Indexing is an ongoing process of fine-tuning and continuous improvement, but the index advisor is a tool that can help get you started. Indexing is almost never "perfect" or "done", so keep tweaking until you're satisfied.

For a complete overview of indexing in Couchbase, make sure to visit the link:https://docs.couchbase.com/server/current/learn/services-and-indexes/indexes/indexing-and-query-perf.html[Indexing] documentation.

If you're still struggling with indexing, you can always get help from humans on the link:https://forums.couchbase.com/c/n1ql[N1QL category on the Couchbase Forums].

Return to the parent User Profile tutorial:

> * link:http://docs.couchbase.com/tutorials/profile-store/dotnet.html[.NET]

Reviewer: ```suggestion
* xref:dotnet.adoc[.NET]
```

> * link:http://docs.couchbase.com/tutorials/profile-store/java.html[Java]

Reviewer: ```suggestion
* xref:java.adoc[Java]
```
